extmod/uasyncio/core.py

```diff
@@ -53,7 +53,10 @@ def __next__(self):
 # Use a SingletonGenerator to do it without allocating on the heap
 def sleep_ms(t, sgen=SingletonGenerator()):
     assert sgen.state is None
-    sgen.state = ticks_add(ticks(), max(0, t))
+    t = max(0, t)
+    if ticks_diff(t, 0) != t:
```

Comment:
I'm not sure this is 100% correct.  `ticks_diff` assumes both args are results from `ticks_ms()` (or `ticks_us()`) and are essentially opaque values (as an implementation they are small ints).  So this may in some rare cases accidentally return `t` even if `t` is too large.

Maybe use `ticks_add(0, t)`?  At least that will raise an exception if `t` is larger than a machine word in size.

Reply:
> I'm not sure this is 100% correct. `ticks_diff` assumes both args are results from `ticks_ms()` (or `ticks_us()`) and are essentially opaque values (as an implementation they are small ints). So this may in some rare cases accidentally return `t` even if `t` is too large.

Related to this, I don't think `ticks_diff` should make this assumption and should instead use `mp_obj_get_int`. If this assumption isn't true, the behaviour is very wrong.

> So this may in some rare cases accidentally return t even if t is too large.

Can you provide an example? Because the result goes via `% MICROPY_PY_UTIME_TICKS_PERIOD` ticks_diff can't return anything outside the range `(-MICROPY_PY_UTIME_TICKS_PERIOD/2, MICROPY_PY_UTIME_TICKS_PERIOD/2)`. So a "too large" value of `t` cannot be equal.

Reply:
You should only ever pass the output of `ticks_ms/us` to `ticks_diff` and `ticks_add`.  More formally:
```py
ticks_ms() -> T
ticks_us() -> T
ticks_diff(T, T) -> int
ticks_add(T, int) -> T  # and may raise an exception if int is too large
```
You should consider `T` as an opaque type that you know nothing about.

> Can you provide an example?

No, I don't think I can!  I was thinking on some machines the address of the big-int would be larger than the value and then wrap around... but as a strange example of what you can get:
```py
big=1<<70
zero = big - big  # a big zero!
print(hex(time.ticks_diff(zero,0)))
print(hex(time.ticks_diff(zero+123456,0)))
```
This will give varying results on 32/64/nanbox builds.  Eg:
```
0x3f9db9f30020
0x3f9db9f30760
```

Reply:
> You should consider T as an opaque type that you know nothing about.

Sure, but this is a trap for people who haven't carefully read and understood every line of the documentation. This is pretty subtle!

>  but as a strange example of what you can get:

This is kind of my point -- it's a dangerous optimisation. Clearly `time.ticks_diff(<thing that is zero>,0)` should be zero.

But either way, it doesn't return 0, so for the purpose of this PR `ticks_diff` is doing the right thing?

Reply:
> But either way, it doesn't return 0, so for the purpose of this PR ticks_diff is doing the right thing?

Not really, because code like `await asyncio.sleep(big_zero)` will now fail, when it should work.